pkg/roachprod/blobfixture/registry.go

```diff
@@ -0,0 +1,282 @@
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package blobfixture
+
+import (
+	"context"
+	"net/url"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/cockroach/pkg/cloud/cloudpb"
+	"github.com/cockroachdb/cockroach/pkg/cloud/nodelocal"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/logger"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/ioctx"
+	"github.com/cockroachdb/errors"
+)
+
+// Registry is used to manage fixtures stored in object storage. It tracks
+// which fixtures are ready for use and uses garbage collection to delete
+// leaked clusters.
+type Registry struct {
+	// storage is used to read and write fixture data and metadata. storage is
+	// configured using the root of the bucket. For example, if the uri is
+	// gs://cockroach-fixtures/roachprod/v25.1, the 'roachprod/v25.1' part is
+	// removed when creating the storage instance. This improves the legibility
+	// of paths in fixture metadata objects.
+	storage cloud.ExternalStorage
+
+	// clock is used to generate timestamps for fixtures. It is a variable so that
+	// it can be replaced in tests.
+	clock func() time.Time
+
+	// uri is the prefix for all fixture data and metadata.
+	//
+	// Fixtures are stored in the following layout:
+	// <uri>/metadata/<kind>/<timestamp> contains metadata for a fixture instance
+	// <uri>/<kind>/<timestamp> contains the actual fixture data
+	//
+	// The uri ispassed to the registry at construction time. The baseURI but it
+	// is expected to be of the form "scheme://<bucket>/roachprod/<version>". So
+	// a full metadata path looks like:
+	// gs://cockroach-fixtures/roachprod/v25.1/metadata/backup-tpcc-30k/20220101-1504
+	uri url.URL
+}
+
+// NewRegistry creates a new Registry instance. The uri parameter is the prefix
+// for all fixture data and metadata. See the comment on the uri field for the
+// structure of a fixture directory.
+func NewRegistry(ctx context.Context, uri url.URL) (*Registry, error) {
+	supportedSchemes := map[string]bool{"gs": true, "s3": true, "azure": true, "nodelocal": true}
+	if !supportedSchemes[uri.Scheme] {
+		return nil, errors.Errorf("unsupported scheme %q", uri.Scheme)
+	}
+
+	baseUri := uri
+	baseUri.Path = ""
+
+	var storage cloud.ExternalStorage
+	var err error
+	if uri.Scheme != "nodelocal" {
+		storage, err = cloud.EarlyBootExternalStorageFromURI(
+			ctx,
+			baseUri.String(),
+			base.ExternalIODirConfig{},
+			cluster.MakeTestingClusterSettings(),
+			nil, /*limiters */
+			cloud.NilMetrics,
+		)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// 'nodelocal://' support is a little hacky and is only intended to support
+		// testing fixture generation in local tests.
+		node := uri.Host
+		root := path.Join(os.Getenv("HOME"), "local", node, "data", "extern")
```

Comment:
i assume you used this in demo? It may be worth adding a baby unit test to ensure this works.